File: lld_game_design_primer/minesweeper/Approach_1/Game.py
from Board import Board, GameState


class Game:
    def __init__(self, boardRows, boardColumns, numberOfBombs):
        self._board = None
        self._boardRows = boardRows
        self._boardColumns = boardColumns
        self._numberOfBombsInTheGame = numberOfBombs
    def _initialize(self):
        if self._board is None:
            self._board = Board(self._boardRows, self._boardColumns, self._numberOfBombsInTheGame)
            self._board.printBoard()
    def start(self):
        if self._board is None:
            self._initialize()
        self._playGame()
    def _playGame(self):
        scanner = input()
        gameState = GameState.INPROGRESS
        while gameState is GameState.INPROGRESS:
            inputi = input()
            if inputi == "exit":
                return
            row = int(inputi.split("-")[0])
            col = int(inputi.split("-")[1])
            gameState = self._board.exposeCell(row, col)
            if gameState is GameState.LOST:
                print("FAIL")
            elif gameState is GameState.WON:
                print("WIN")
            self._board.printBoard()
    

    

File: lld_game_design_primer/minesweeper/Approach_1/Game.java
package lld_game_design_primer.minesweeper.Approach_1;

import java.util.Random;
import java.util.Scanner;


public class Game {

    private Board board;

    private int boardRows;
    private int boardColumns;
    private int numberOfBombsInTheGame;
   

    public Game(int boardRows, int boardColumns, int numberOfBombs) {
        this.boardRows = boardRows;
        this.boardColumns =  boardColumns;
        this.numberOfBombsInTheGame = numberOfBombs;
    }

    private void initialize() {
        if (board == null) {
            board = new Board(boardRows, boardColumns, numberOfBombsInTheGame);
            board.printBoard(); // display board
        }
    }

    public void start() {
        if (board == null) {
            initialize();
        }
        playGame();
    }

    private void playGame() {
        Scanner scanner = new Scanner(System.in);
        Board.GameState gameState = Board.GameState.INPROGRESS;

        while (gameState == Board.GameState.INPROGRESS) {
            String input = scanner.nextLine(); // input the coordinate of the cell you want to click in the form of row-col
            if (input.equals("exit")) { // exit is the keyword to exit the game
                scanner.close();
                return;
            }

            int row = Integer.parseInt(input.split("-")[0]);
            int col = Integer.parseInt(input.split("-")[1]);

            gameState = board.exposeCell(row, col);

            // check if the player won or lost by making the above move
            if (gameState == Board.GameState.LOST) {
                System.out.println("FAIL");
            } else if (gameState == Board.GameState.WON) {
                System.out.println("WIN");
            }

            board.printBoard(); // display board after the player has made the move
        }
        scanner.close();
    }
     
}
